tempesta_fw/cfg.c

```diff
@@ -0,0 +1,1563 @@
+ *  It consists of entries. Each entry has:
+ *    1. name;
+ *    2. values (usually just one, but variable number of values is supported);
+ *    3. attributes (a dictionary of key-value pairs);
+ *    4. children entries (such entries act as sections or trees);
+ *  The only name is required. Everything else is optional. The idea is similar
+ *  to SDL (http://www.ikayzo.org/display/SDL/Language+Guide), but our syntax
+ *  and terminology is more habitual for C/Linux programmers and users.
+ *
+ *  Tempesta FW modules register themselves and provide their configuration
+ *  specifications via TfwCfgMod and TfwCfgSpec structures. The code here pushes
+ *  events and parsed configuration via callbacks specified in these structures.
+ *
+ *  The code in this unit contains two main entities:
+ *    1. The configuration parser.
+ *       We utilize FSM approach for the parser. The code is divided into two
+ *       FSMs: TFSM (tokenizer) and PFSM (the parser that produces entries).
+ *    2. A bunch of generic  TfwCfgSpec->handler callbacks for the parser.
+ *    3. TfwCfgMod list related routines, the top-level parsing routine.
+ *       This part of code implements publishing start/stop events and parsed
+ *       configuration data across modules.
+ *    4. The list of registered modules, VFS and sysctl helpers, kernel module
+ *       parameters. The stateful part of code.
+ *
+ * TODO:
+ *  - "include" directives.
+ *  - Handling large sets of data, possibly via TDB.
+ *  - Re-loading some parts of the configuration on the fly without stopping the
+ *    whole system.
+ *  - Verbose error reporting: include file/line and expected/got messages.
+ *  - Improve efficiency: too many memory allocations and data copying.
+ *
+ * Copyright (C) 2012-2014 NatSys Lab. (info@natsys-lab.com).
+ * Copyright (C) 2015 Tempesta Technologies.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#include <linux/ctype.h>
+#include <linux/kernel.h>
+#include <linux/moduleparam.h>
+#include <net/net_namespace.h> /* for sysctl */
+
+#include "addr.h"
+#include "lib.h"
+#include "log.h"
+
+#include "cfg.h"
+
+/* FSM's debug messages are very verbose, so they are turned off by default. */
+#ifdef DEBUG_CFG_FSM
+#define FSM_DBG(...) TFW_DBG(__VA_ARGS__)
+#else
+#define FSM_DBG(...)
+#endif
+
```

Comment:
> 1. prepend mudule specific line to the logging message, like "cfg" for this case, so they can differentiate messages from different modules;

Yea, but then I have to grep logs every time, and do that in a tricky way: I have to remove unwanted messages instead of enabled messages.

Suppose I want to investigate the "cfg" module. Then I need a log containing debug messages only for the "cfg" module, plus regular non-debug messages from all modules. Then I can't simply do `grep cfg` because it wipes out messages from other modules, and I can't simply do `grep -v mod1 | grep -v mod2 | etc` because it filters out non-debug messages. 

I know that could be handled by tricky message formatting and clever scripts, but for me that looks unhandy at the moment.

> 2. use only one environment/build variable with debug level. Will you create 5 different variables and how will you name them if you need 5 different levels of debug messages?

Yes, there will be 5 different entities (not variables but compile-time switches).

But you can't avoid that. If you need to distinguish messages, then you have to name them somehow. Perhaps you choose to prepend some prefixes to messages (like "tfw_cfg: blah" or "tfw_sched_http: blah") and then `grep` them. That actually doesn't make much difference since you have to create 5 prefixes and agree on their naming. 

Compile-time switches look handy for me. Debugging is a part of every-day programming process. I work with the source code all the time, so for me it is worthless to change a single line and run Tempesta FW again, and get a readable text log without additional scripts. I don't want to complicate this process.

Actually this way is only applicable if you can reproduce a bug easily on your machine. Things go complicated when you met a heisenbug or you have to use a remote setup to reproduce the bug. I suggest to postpone the work until such case.